chapter_recurrent-neural-networks/sequence.ipynb

```diff
@@ -0,0 +1,584 @@
+    "NDArray x = time.mul(0.01).sin().add(\n",
+    "    manager.randomNormal(0f, 0.2f, new Shape(T), DataType.FLOAT32));\n",
+    "\n",
+    "double[] xAxis = Functions.floatToDoubleArray(time.toFloatArray());\n",
+    "double[] yAxis = Functions.floatToDoubleArray(x.toFloatArray());\n",
+    "\n",
+    "plot(xAxis, yAxis, \"time\", \"x\");"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "origin_pos": 6
+   },
+   "source": [
+    "Next, we need to turn such a sequence into features and labels that our model can train on.\n",
+    "Based on the embedding dimension $\\tau$ we map the data into pairs $y_t = x_t$ and $\\mathbf{x}_t = [x_{t-\\tau}, \\ldots, x_{t-1}]$.\n",
+    "The astute reader might have noticed that this gives us $\\tau$ fewer data examples, since we do not have sufficient history for the first $\\tau$ of them.\n",
+    "A simple fix, in particular if the sequence is long,\n",
+    "is to discard those few terms.\n",
+    "Alternatively we could pad the sequence with zeros.\n",
+    "Here we only use the first 600 feature-label pairs for training.\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "int tau = 4;\n",
+    "NDArray features = manager.zeros(new Shape(T - tau, tau));\n",
+    "\n",
+    "for (int i = 0; i < tau; i++) {\n",
+    "    features.set(new NDIndex(\":, \" + i), x.get(new NDIndex(\"\" + i + \":\" + (T - tau + i) )));\n",
+    "}\n",
+    "NDArray labels = x.get(new NDIndex(\"\" + tau + \":\")).reshape(new Shape(-1,1));"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "int batchSize = 16;\n",
+    "int nTrain = 600;\n",
+    "// Only the first `nTrain` examples are used for training\n",
+    "ArrayDataset trainIter = new ArrayDataset.Builder()\n",
+    "    .setData(features.get(new NDIndex(\":\" + nTrain)))\n",
+    "    .optLabels(labels.get(new NDIndex(\":\" + nTrain)))\n",
+    "    .setSampling(batchSize, false)\n",
+    "    .build()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "origin_pos": 10
+   },
+   "source": [
+    "Here we keep the architecture fairly simple:\n",
+    "just an MLP with two fully-connected layers, ReLU activation, and squared loss.\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "// A simple MLP\n",
+    "public SequentialBlock getNet() {\n",
+    "    SequentialBlock net = new SequentialBlock();\n",
+    "//     net.add(Blocks.batchFlattenBlock(4));\n",
+    "    net.add(Linear.builder().setUnits(10).build());\n",
+    "    net.add(Activation::relu);\n",
+    "    net.add(Linear.builder().setUnits(1).build());\n",
+    "    \n",
+    "    net.setInitializer(new XavierInitializer());\n",
+    "    net.initialize(manager, DataType.FLOAT32, new Shape(4));\n",
```

Comment:
you can remove setInitializer & initialize as we do this when we use Trainer

Reply:
Done